File: euler_037.erl
-module(euler_037).
-export([solve/0]).
-import(primes).

% The number 3797 has an interesting property. Being prime itself, it is
% possible to continuously remove digits from left to right, and remain
% prime at each stage: 3797, 797, 97, and 7. Similarly we can work from
% right to left: 3797, 379, 37, and 3.
%
% Find the sum of the only eleven primes that are both truncatable from left
% to right and right to left.
%
% NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.

solve() ->
    T1 = now(),
    Count = start(),
    T2 = now(),
    Time = timer:now_diff(T2,T1)/1000,
    io:format(
        "The sum of all 11 truncatable primes is ~p (~p ms) ~n",
        [Count, Time]
    ).

start() ->
    Primes = primes:generate(1000000),
    Trunc = [ T ||
        T <- Primes,
        is_truncatable(T)
    ],
    lists:foldl(
        fun(X, Acc) ->
            Acc + X
        end, 0, Trunc
    ).

is_truncatable(T) when T < 10 ->
    false;
is_truncatable(T) ->
    B = integer_to_list(T),
    trunc_l(B) and trunc_r(B).

trunc_l(T) when length(T) == 1 ->
    primes:is_prime(list_to_integer(T));
trunc_l(List) ->
    [_|T] = List,
    case primes:is_prime(list_to_integer(T)) of
        true -> trunc_l(T);
        false -> false
    end.

trunc_r(T) when length(T) == 1 ->
    primes:is_prime(list_to_integer(T));
trunc_r(List) ->
    {H, _} = lists:split(
        length(List)-1, List
    ),
    case primes:is_prime(list_to_integer(H)) of
        true -> trunc_r(H);
        false -> false
    end.

File: euler_037.rb
require "./primes.rb"

# The number 3797 has an interesting property. Being prime itself, it is
# possible to continuously remove digits from left to right, and remain
# prime at each stage: 3797, 797, 97, and 7. Similarly we can work from
# right to left: 3797, 379, 37, and 3.
#
# Find the sum of the only eleven primes that are both truncatable from left
# to right and right to left.
#
# NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.

class Euler_037

    def initialize
        printf(
            "=== Euler 037 ============= \n"
        )
    end

    def solve
        start = Time.now
        primes = Primes.generate 1000000
        answer = truncatable primes
        stop = Time.now
        printf(
            "The sum of all 11 truncatable primes is: %d (%.4f ms.)\n",
            answer,
            ((stop - start) * 1000)
        )
    end

    def truncatable list
        count = 0
        answer = 0
        list.each do |x|
            next if x < 10
            break if count == 11
            if truncatable_prime? x
                count += 1
                answer += x
            end
        end
        return answer
    end

    def truncatable_prime? i
        nums = i.to_s.chars.to_a
        nums.length.times do
            return false if (!Primes.prime? nums.join().to_i)
            nums.shift
        end

        nums = i.to_s
        while nums.length > 0
            return false if (!Primes.prime? nums.to_i)
            nums.chop!()
        end
        return true
    end
end


Euler_037.new.solve